### PR TITLE
Maven integration

### PR DESCRIPTION
**Pull request by fcrespo82**
Opened this pull request for discussion on where to improve and fixes to do on this feature before merging into master if approved.

### CODE REVIEW

**Comment by pflannery** [comment]
@fcrespo82 Thanks for the PR. 

Before I can start to looking at this please can you convert all .ts files to .js files because this isn't a typescript project. It only uses typescript to compile.

Can you also provide a smoke test in a new folder called maven under https://github.com/vscode-contrib/vscode-versionlens/tree/master/test/smoke

**Comment by fcrespo82** [comment]
Of course... I'll do that!

There is still work to be done before the merge... 

I'll let you know when is ready!

**Comment by fcrespo82** [comment]
Hey @pflannery. I think the pull request is good enough for merging.

Can you take a look and point me if there's anything else I can improve.

Thanks for your time!

**Comment by pflannery** [comment]
@fcrespo82 After you resolve my comments above can you also merge the lastest master in to this PR and resolve the conflicts.

**Comment by fcrespo82** [comment]
Just merged back. All the other concerns were already resolved.
Can you check back?

**Comment by fcrespo82** [comment]
Just merged back. All the other concerns were already resolved.
Can you check back?

**Comment by fcrespo82** [comment]
Sorry, I missed that... Of course I'll do it..

**Comment by pflannery** [comment]
@fcrespo82 It's looking good so far however I've noticed that the current selected version analysis is appearing in different places. 

image

Please make sure it always appears first in the list for each entry

image

**Comment by fcrespo82** [comment]
Hi! I fixed that inconsistency and fixed one with tagged versions too.

Should be better now..

**Comment by pflannery** [comment]
@fcrespo82 Everything is great. 

I noticed that you introduced a `wrapVersion` function to try and get around not having to include speech marks for child xml tags `<version>#.#.#</version>`

When this project was first setup up the json parser returned start and end offsets that included the speech marks. So when the dotnet parser came along it was implemented in the same manor but it doesn't actually have child xml `<version>` tags so we never hit the issue you found until now.

So as this project grows I think it was best that I update the json and xml parsers to exclude speech marks from the start and end offsets.

This now means that you don't need to include a `wrapVersion` in your PR anymore but it does mean there is a conflict now (shouldn't be too hard to resolve).

File: src/common/packageCodeLens.js [hunk 1]
```diff
@@ -24,6 +24,12 @@ export class PackageCodeLens extends CodeLens {
     return this.package.customGenerateVersion.call(this, this.package, newVersion);
   }
 
+  wrapVersion(version) {
```
Comment (pflannery) [inline src/common/packageCodeLens.js]: You can remove wrapVersion

File: src/commands/factory.js [hunk 2]
```diff
@@ -114,7 +114,7 @@ export function createGithubCommand(codeLens) {
       return codeLens.setCommand(
         `${meta.category}: ${codeLens.getInstallIndicator()} ${entry.version}`,
         `${appSettings.extensionName}.updateDependencyCommand`,
-        [codeLens, `"${newVersion}"`]
+        [codeLens, `${codeLens.wrapVersion(newVersion)}`]
```
Comment (pflannery) [inline src/commands/factory.js]: Undo your edit wrapVersion here

File: src/commands/factory.js [hunk 3]
```diff
@@ -71,7 +71,7 @@ export function createSatisfiedWithNewerCommand(serverVersion, codeLens) {
   return codeLens.setCommand(
     `Satisfies ${codeLens.getInstallIndicator()} ${serverVersion}`,
     `${appSettings.extensionName}.updateDependencyCommand`,
-    [codeLens, `"${replaceWithVersion}"`]
+    [codeLens, `${codeLens.wrapVersion(replaceWithVersion)}`]
```
Comment (pflannery) [inline src/commands/factory.js]: Undo your edit wrapVersion here

File: src/commands/factory.js [hunk 4]
```diff
@@ -58,7 +58,7 @@ export function createNewVersionCommand(newVersion, codeLens) {
   return codeLens.setCommand(
     `${codeLens.getTaggedVersionPrefix()}${codeLens.getInstallIndicator()} ${newVersion}`,
     `${appSettings.extensionName}.updateDependencyCommand`,
-    [codeLens, `"${replaceWithVersion}"`]
+    [codeLens, `${codeLens.wrapVersion(replaceWithVersion)}`]
```
Comment (pflannery) [inline src/commands/factory.js]: Undo your edit wrapVersion here

File: src/common/packageGeneration.js [hunk 5]
```diff
@@ -43,11 +43,12 @@ export function createInvalidVersion(name, version, type) {
   );
 }
 
-export function createPackage(name, version, meta, customGenerateVersion) {
+export function createPackage(name, version, meta, customGenerateVersion, customWrapVersion) {
```
Comment (pflannery) [inline src/common/packageGeneration.js]: remove customWrapVersion here

File: src/common/packageGeneration.js [hunk 6]
```diff
@@ -43,11 +43,12 @@ export function createInvalidVersion(name, version, type) {
...
   return {
     name,
     version,
     meta,
-    customGenerateVersion
+    customGenerateVersion,
+    customWrapVersion
```
Comment (pflannery) [inline src/common/packageGeneration.js]: remove customWrapVersion here

File: src/providers/maven/mavenPackageParser.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import * as PackageFactory from 'common/packageGeneration';
+import { buildMapFromVersionList, buildTagsFromVersionMap } from './versionUtils'
+
+export function mavenPackageParser(name, requestedVersion, appContrib) {
+
+  // get all the versions for the package
+  return mavenGetPackageVersions(name)
+    .then((versions) => {
+      // console.log(versions);
+      let customWrapVersion = (v) => {
```
Comment (pflannery) [inline src/providers/maven/mavenPackageParser.js]: remove customWrapVersion here

File: src/providers/maven/mavenPackageParser.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            type: 'maven',
+            tag
+          };
+
+          return PackageFactory.createPackage(
+            name,
+            requestedVersion,
+            meta,
+            null,
+            customWrapVersion
```
Comment (pflannery) [inline src/providers/maven/mavenPackageParser.js]: remove customWrapVersion here

**Comment by fcrespo82** [comment]
Hi @pflannery I updated the branch removing the wrapVersion, added the support to use the repositories set on the current pom and also fixed the regex for tagged version and if the url didn't ends with "/". I updated the smoke test and should be good to go.

My colleagues at work are excited for this PR. They found it great for Java development.

Thanks again for the awesome extension.

**Pull request merged**